src/workerd/io/actor-sqlite.c++

```diff
@@ -331,11 +452,52 @@ void ActorSqlite::shutdown(kj::Maybe<const kj::Exception&> maybeException) {
 }
 
 kj::Maybe<kj::Own<void>> ActorSqlite::armAlarmHandler(kj::Date scheduledTime, bool noCache) {
-  return hooks.armAlarmHandler(scheduledTime, noCache);
+  ensureAlarmTimeInitialized();
+
+  KJ_ASSERT(!currentAlarmTime.is<DeferredAlarmDelete>());
+
+  bool alarmDeleteNeeded = true;
+  KJ_IF_SOME(t, currentAlarmTime.tryGet<KnownAlarmTime>()) {
+    if (t.time != scheduledTime) {
+      if (t.status == KnownAlarmTime::Status::CLEAN) {
+        // If there's a clean scheduledTime that is different from ours, this run should be
+        // canceled.
+        return kj::none;
+      } else {
+        // There's a alarm write that hasn't been set yet pending for a time different than ours --
+        // We won't cancel the alarm because it hasn't been confirmed, but we shouldn't delete
+        // the pending write.
+        alarmDeleteNeeded = false;
+      }
+    }
+  }
+
+  // TODO(now): Here, similar to the ActorCache code, we're assuming that if the alarm time
+  // matches but is not in a clean state (i.e. is dirty or flushing), that it's OK to schedule the
+  // delete, and that it's OK to use the current alarm time to initialize the alarm state back to
+  // "clean" if the alarm is cancelled.  Is that OK?
```

Comment:
It's apparently not disastrous given that we've been relying on it for quite some time now. I'd allow it, since it seems pretty likely that if we're getting invoked at the given time that the time was probably successfully flushed.

I suppose that the sqlite storage does add a second source of truth that makes this a little more confusing though. But even so, if the alarm time matched, and we successfully ran an alarm invocation at that time, then it seems correct to delete it?

So I think this is fine, but let me know if you see an issue that I'm missing.

Reply:
Re: the sqlite storage adding a second source of truth... My original intent was for the sqlite storage to be the source of truth, at least for the local, possibly-not-yet-persisted alarm state, but I wasn't sure if it would be expensive to query sqlite each time we need to check the alarm value, so I'm essentially caching it in the `currentAlarmTime` member.  If it's actually not that expensive to run the query, I think I could make the code a bit simpler.

Reply:
I'd think it's fine to just hit sqlite every time. I don't know what @kentonv thinks.

It could have a very minor affect on billing since it would mean getAlarm() counts as one row read, but (A) I think we're going technically going to start causing a row to be read by the first getAlarm even with this caching, and (B) rows read are so preposterously dirt cheap that it's hard to imagine it becoming a big issue. That'll also depend on how #2707 and the related work plays out. I believe the goal was to not start charging for getAlarm, which may require some extra care to achieve :/

Reply:
With the current implementation of counting rowsRead/rowsWritten, alarms will also be counted. Could you open a ticket under https://jira.cfdata.org/browse/STOR-3551 as a bug so we don't loose track?

Reply:
https://jira.cfdata.org/browse/STOR-3744 - will need to be fixed before we enable usage based billing

Reply:
I think it's fine to query sqlite again every time and I think it's fine that this is billed as a row read.

Reply:
Hi :wave:,

About half a year ago on Discord it was confirmed by @a-robinson that getAlarm would not be billed: https://discord.com/channels/595317990191398933/773219443911819284/1232360354210648136 - so (without access to the tickets and without knowing your plans) to make sure that people are aware of the change, please include this as a note somewhere in the docs or changelog.

Not knowing pricing (just a guess based on D1) right now, I'm not worried about sudden increases in bills, I'm just asking for this to be documented, please 😁